sdk/queue/priority_queue.go

```diff
@@ -0,0 +1,187 @@
+	// Key is a unique string used to identify items in the internal data map
+	Key string
+	// Value is an unspecified type that implementations can use to store
+	// information
+	Value interface{}
+
+	// Priority determines ordering in the queue, with the lowest value being the
+	// highest priority
+	Priority int64
+
+	// index is an internal value used by the heap package, and should not be
+	// modified by any consumer of the priority queue
+	index int
+}
+
+// NewPriorityQueue initializes the internal data structures and returns a new
+// PriorityQueue
+func NewPriorityQueue() *PriorityQueue {
+	pq := PriorityQueue{
+		data:    make([]*Item, 0),
+		dataMap: make(map[string]*Item),
+	}
+	heap.Init(&pq)
+	return &pq
+}
+
+// PriorityQueue satisfies heap.Interface and sort.Interface, and adds a few
+// convenience methods. The ordering (priority) is an int64 value with the
+// smallest value is the highest priority.
+// See https://golang.org/pkg/container/heap/#example__priorityQueue
+type PriorityQueue struct {
+	// data is the internal structure that holds the queue, and is operated on by
+	// heap functions
+	data []*Item
+
+	// dataMap represents all the items in the queue, with unique indexes, used
+	// for finding specific items. dataMap is kept in sync with the data slice
+	dataMap map[string]*Item
+
+	// mapMutex is used to facilitate read/write locks on the dataMap
+	dataMutex sync.Mutex
+}
+
+// PopItem pops the highest priority item from the queue. This is a
+// wrapper/convenience method that calls heap.Pop, so consumers do not need to
+// invoke heap functions directly
+func (pq *PriorityQueue) PopItem() (*Item, error) {
+	pq.dataMutex.Lock()
+	defer pq.dataMutex.Unlock()
+
+	if pq.Len() == 0 {
+		return nil, ErrEmpty
+	}
+	item := heap.Pop(pq).(*Item)
+	delete(pq.dataMap, item.Key)
+	return item, nil
+}
+
+// PushItem pushes an item on to the queue. This is a wrapper/convenience
+// method that calls heap.Push, so consumers do not need to invoke heap
+// functions directly. Items must have unique Keys, and Items in the queue
+// cannot be updated. To modify an Item, users must first remove it and re-push
+// it after modifications
+func (pq *PriorityQueue) PushItem(i *Item) error {
+	if i.Key == "" {
+		return errors.New("error adding item: Item Key is required")
+	}
+
+	pq.dataMutex.Lock()
+	defer pq.dataMutex.Unlock()
+
+	if _, ok := pq.dataMap[i.Key]; ok {
+		return ErrDuplicateItem
+	}
+
+	// copy the item value(s) so that modifications to the source item does not
+	// affect the item on the queue
+	clone, err := copystructure.Copy(i)
+	if err != nil {
+		return err
+	}
+
+	pq.dataMap[i.Key] = clone.(*Item)
+	heap.Push(pq, clone)
+	return nil
+}
+
+// PopItemByKey searches the queue for an item with the given key and removes it
+// from the queue if found. Returns ErrItemNotFound(key) if not found. This
+// method must fix the queue after removal.
+func (pq *PriorityQueue) PopItemByKey(key string) (*Item, error) {
```

Comment:
Can we rename this to `Delete` since it is more of a map like function?

Reply:
I prefer `Pop...` because it's consistent with the idea of operating on a queue, in my opinion. If others feel strongly about it I am willing to reconsider but for now I'm going to leave it as is. Thanks for the feedback

Reply:
`PopByKey` seems like an ok compromise.

Reply:
👍